internal/impl/kafka/enterprise/redpanda_migrator_offsets_input.go

```diff
@@ -0,0 +1,207 @@
+				Advanced().
+				Default("1MiB"),
+		},
+		kafka.FranzReaderOrderedConfigFields(),
+		[]*service.ConfigField{
+			service.NewAutoRetryNacksToggleField(),
+		},
+	)
+}
+
+func init() {
+	err := service.RegisterBatchInput("redpanda_migrator_offsets", redpandaMigratorOffsetsInputConfig(),
+		func(conf *service.ParsedConfig, mgr *service.Resources) (service.BatchInput, error) {
+			tmpOpts, err := kafka.FranzConnectionOptsFromConfig(conf, mgr.Logger())
+			if err != nil {
+				return nil, err
+			}
+			clientOpts := append([]kgo.Opt{}, tmpOpts...)
+
+			d := kafka.FranzConsumerDetails{}
+
+			var topics []string
+			if topicList, err := conf.FieldStringList(rmoiFieldTopics); err != nil {
+				return nil, err
+			} else {
+				topics, _, err = kafka.ParseTopics(topicList, -1, false)
+				if err != nil {
+					return nil, err
+				}
+				if len(topics) == 0 {
+					return nil, errors.New("at least one topic must be specified")
+				}
+			}
+
+			var topicPatterns []*regexp.Regexp
+			if regexpTopics, err := conf.FieldBool(rmoiFieldRegexpTopics); err != nil {
+				return nil, err
+			} else if regexpTopics {
+				topicPatterns = make([]*regexp.Regexp, 0, len(topics))
+				for _, topic := range topics {
+					tp, err := regexp.Compile(topic)
+					if err != nil {
+						return nil, fmt.Errorf("failed to compile topic regex %q: %s", topic, err)
+					}
+					topicPatterns = append(topicPatterns, tp)
+				}
+			}
+
+			if d.RackID, err = conf.FieldString(rmoiFieldRackID); err != nil {
+				return nil, err
+			}
+
+			if d.FetchMaxBytes, err = kafka.BytesFromStrFieldAsInt32(rmoiFieldFetchMaxBytes, conf); err != nil {
+				return nil, err
+			}
+			if d.FetchMinBytes, err = kafka.BytesFromStrFieldAsInt32(rmoiFieldFetchMinBytes, conf); err != nil {
+				return nil, err
+			}
+			if d.FetchMaxPartitionBytes, err = kafka.BytesFromStrFieldAsInt32(rmoiFieldFetchMaxPartitionBytes, conf); err != nil {
+				return nil, err
+			}
+
+			// Consume messages from the `__consumer_offsets` topic
+			d.Topics = []string{`__consumer_offsets`}
+			clientOpts = append(clientOpts, d.FranzOpts()...)
+
+			rdr, err := kafka.NewFranzReaderOrderedFromConfig(conf, mgr, func() ([]kgo.Opt, error) {
+				return clientOpts, nil
+			}, func(record *kgo.Record) (*service.Message, error) {
+				msg := kafka.FranzRecordToMessageV1(record)
+
+				// Check the version to ensure that we process only offset commit keys
+				key := kmsg.NewOffsetCommitKey()
+				if err := key.ReadFrom(record.Key); err != nil || (key.Version != 0 && key.Version != 1) {
+					return nil, fmt.Errorf("failed to decode record key: %s", err)
+				}
+
+				isExpectedTopic := false
+				if len(topicPatterns) > 0 {
```

Comment:
Can we instead pull a predicate out as a local variable so this becomes: `isExpectedTopic := matchesTopic(key.Topic)` or something like that?

Reply:
I took a shot at it, but it I'm not sure if that's what you had in mind. Let me know if this looks OK.

I think it would be nice to introduce a generic helper similar to https://github.com/redpanda-data/benthos/blob/9fdcbd09cc5c8b8276728e14a18a4612b7e0460c/internal/metadata/include_filter.go#L75, but I'll leave that for when we'll have more use cases.